src/segmented_cache_server.erl

```diff
@@ -0,0 +1,156 @@
+    true = is_integer(N) andalso N > 0,
+    TTL0 = maps:get(ttl, Opts, {hours, 8}),
+    TTL = case TTL0 of
+               infinity -> infinity;
+               {milliseconds, S} -> S;
+               {seconds, S} -> timer:seconds(S);
+               {minutes, M} -> timer:minutes(M);
+               {hours, H} -> timer:hours(H);
+               T when is_integer(T) -> timer:minutes(T)
+           end,
+    true = (TTL =:= infinity) orelse (is_integer(TTL) andalso N > 0),
+    Strategy = maps:get(strategy, Opts, fifo),
+    true = (Strategy =:= fifo) orelse (Strategy =:= lru),
+    MergerFun = maps:get(merger_fun, Opts, fun ?MODULE:default_merger_fun/2),
+    true = is_function(MergerFun, 2),
+    Scope = maps:get(scope, Opts, pg),
+    true = (undefined =/= whereis(Scope)),
+    {Scope, N, TTL, Strategy, MergerFun}.
+
+-ifdef(OTP_RELEASE).
+  -if(?OTP_RELEASE >= 25).
+    ets_settings() ->
+        [set, public,
+         {read_concurrency, true},
+         {write_concurrency, auto},
+         {decentralized_counters, true}].
+  -elif(?OTP_RELEASE >= 21).
+    ets_settings() ->
+        [set, public,
+         {read_concurrency, true},
+         {write_concurrency, true},
+         {decentralized_counters, true}].
+  -endif.
+-else.
+ets_settings() ->
+    [set, public,
+     {read_concurrency, true},
+     {write_concurrency, true},
+     {decentralized_counters, true}].
+-endif.
+
+-spec handle_call(any(), gen_server:from(), #cache_state{}) -> {reply, ok, #cache_state{}}.
+handle_call(_Msg, _From, State) ->
+    {reply, ok, State}.
+
+-spec handle_cast(term(), #cache_state{}) -> {noreply, #cache_state{}}.
+handle_cast({delete_entry, Key}, #cache_state{name = Name} = State) ->
+    do_delete(Name, Key, entry, fun ?HELPERS:delete_entry_fun/2),
+    {noreply, State};
+handle_cast({delete_pattern, Pattern}, #cache_state{name = Name} = State) ->
+    do_delete(Name, Pattern, pattern, fun ?HELPERS:delete_pattern_fun/2),
+    {noreply, State};
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+
+-spec do_delete(segmented_cache:name(), term(), entry | pattern, fun()) -> term().
+do_delete(Name, Value, Type, Fun) ->
+    try
+        ?HELPERS:iterate_fun_in_tables(Name, Value, Fun)
+    catch Class:Reason ->
+              Metadata = #{name => Name, delete_type => Type, value => Value, class => Class, reason => Reason},
+              telemetry:execute([segmented_cache, Name, delete_error], #{}, Metadata)
+    end.
+
+-spec handle_info(any(), #cache_state{}) -> {noreply, #cache_state{}}.
+handle_info(purge, #cache_state{ttl = TTL} = State) ->
+    purge_last_segment_and_rotate(State),
+    case TTL of
+        infinity -> {noreply, State};
+        _ -> {noreply, State#cache_state{timer_ref = erlang:send_after(TTL, self(), purge)}}
+    end;
+handle_info(_Msg, State) ->
+    {noreply, State}.
+
+terminate(_Reason, #cache_state{scope = Scope, name = Name, timer_ref = TimerRef}) ->
+    pg:leave(Scope, Name, self()),
```

Comment:
there's no need for manual `pg:leave/3` call, see https://www.erlang.org/doc/man/pg.html#description

```
If a member terminates, it is automatically removed from the group.
```